### ISSUE TITLE
multithreads in http

### ISSUE THREAD

**Issue by suxunbin** [eidheim/Simple-Web-Server#220]
I find a problem when I use server_http.hpp(old version) in http.
![qq 20180418201558](https://user-images.githubusercontent.com/22476651/38931524-0277b97e-4303-11e8-9c8e-ecd4d276770e.png)
"response" is a shared_ptr<HttpServer:: Response> type. In this way, *response can send data to the client.
However,
![qq 20180418201650](https://user-images.githubusercontent.com/22476651/38931639-64551902-4303-11e8-81b2-3ce956b41f42.png)
![qq 20180418202443](https://user-images.githubusercontent.com/22476651/38931707-9f59ef0a-4303-11e8-9999-b58e50a1e48c.jpg)
When I let "Res" equal to "response",*response can not send data to the client.

Why I can't pass  a shared_ptr<HttpServer:: Response> to another.

**Comment by eidheim** [comment]
What is the signature of query_thread?

**Comment by suxunbin** [comment]
![4](https://user-images.githubusercontent.com/22476651/38935660-6a5c1052-430e-11e8-8e29-4e2c0e334c00.png)
![6](https://user-images.githubusercontent.com/22476651/38935668-701c998a-430e-11e8-8026-767935034f73.jpg)
![7](https://user-images.githubusercontent.com/22476651/38935674-73af335a-430e-11e8-95e8-cd8038eaaf75.png)
![8](https://user-images.githubusercontent.com/22476651/38935681-7779a4de-430e-11e8-8096-29858b8d4062.jpg)

![10](https://user-images.githubusercontent.com/22476651/38935707-88d93668-430e-11e8-8818-1b6333122651.png)
![12](https://user-images.githubusercontent.com/22476651/38935723-8de26846-430e-11e8-92bc-4b6bf6237735.png)
Part of ghttp.cpp

**Comment by suxunbin** [comment]
shared_ptr<HttpServer:: Request> is Ok
But
shared_ptr<HttpServer:: Response> is no use when I use pointer assignment

**Comment by suxunbin** [comment]
![1](https://user-images.githubusercontent.com/22476651/38936039-6787159c-430f-11e8-916d-135672be5f4f.jpg)

**Comment by eidheim** [comment]
I could not see anything wrong after a quick look, but making a copy of the `response` shared_ptr should work fine. I'll have another look Tomorrow if you have not figured it out by then.

**Comment by eidheim** [comment]
Ah, the problem seems to be that the Res shared_ptr is kept alive too long. The response is sent when the shared_ptr is destructed, and it seems like your Res object is kept alive in a scope higher than the scope of the handler.

**Comment by suxunbin** [comment]
Yes,I want to store the Res object in order to use it after the handler.
And I find maybe connection between server and client is not broken.
But only *response this ostream doesn't send data to the client.
So make a copy of the response shared_ptr would lead the ostream confusion?

**Comment by suxunbin** [comment]
This is the Response's define in server_http.hpp

![qq 20180419113348](https://user-images.githubusercontent.com/22476651/38969996-9e41287a-4382-11e8-8d4c-5c58b522e132.png)

**Comment by eidheim** [comment]
In that case, you need to use Response::send to send the data you have added to the response stream to the client. See https://github.com/eidheim/Simple-Web-Server/blob/master/http_examples.cpp#L183 for an example.

**Comment by suxunbin** [comment]
[server_http.txt](https://github.com/eidheim/Simple-Web-Server/files/1926884/server_http.txt)

**Comment by suxunbin** [comment]
This is the server_http.hpp that I used.
*response will send the data to the client itself,will it?

**Comment by eidheim** [comment]
Sadly that would be too inefficient. The data is sent when Response::send is called, or when the Response object within the shared_ptr is destroyed.

**Comment by suxunbin** [comment]
Actually,I'm writing a threadpool.
![1](https://user-images.githubusercontent.com/22476651/38979909-10a554de-43ee-11e8-8238-5c27b0569329.PNG)
pool.create() create a daemon thread to continuously get task from tasklines and run query_thread().
Where, tasklines is a global variable
![3](https://user-images.githubusercontent.com/22476651/38979914-142adbe2-43ee-11e8-9aa0-90b0e574f8f6.PNG)
![2](https://user-images.githubusercontent.com/22476651/38979922-16b3e124-43ee-11e8-896c-5acb3a734e3f.PNG)
![4](https://user-images.githubusercontent.com/22476651/38979933-1bee74d8-43ee-11e8-9c5f-55745b93b0ae.PNG)
![5](https://user-images.githubusercontent.com/22476651/38979937-1e8e1888-43ee-11e8-9a99-e0e136f27982.PNG)
However, I find that if I don't sleep(10) in query_handler(), the daemon thread will not get response and request object.
In other words, if I leave query_handler() scope, the daemon thread(detached from the server) will not get response and request object.

**Comment by suxunbin** [comment]
When I am in query_handler() scope,obviously the response and request object are valid wherever they are referred(for example, Task , tasklines, the daemon thread). However, when I leave query_handler() scope, the response and request object are invalid wherever they are referred.
And I want the response and request object to be kept alive in a scope higher than the scope of the handler.(for example , the location of pool.create() )

**Comment by suxunbin** [comment]
![3](https://user-images.githubusercontent.com/22476651/38987510-a9b1cf2c-4402-11e8-9377-0dd256ee9298.PNG)
Can I have a way that when I leave the scope,the response and request object are still alive?
I hope that server.resource doesn't destroy the response and request object when they leave the scope.

**Comment by suxunbin** [comment]
I have read the  https://github.com/eidheim/Simple-Web-Server/blob/master/http_examples.cpp#L183 example.
![6](https://user-images.githubusercontent.com/22476651/38991492-903077c0-4410-11e8-8f4a-2931372f05a4.PNG)
My purpose is to use referrence of the response and request object on the location in the picture.
In other words, I hope to store referrence of the response and request object as global variables in order that other threads can use them.

**Comment by eidheim** [comment]
The purpose of shared_ptr's is typically to extend the lifetime of an object. When you pass the shared_ptr to query_thread, the shared_ptr is copied and thus the object's lifetime is extended. You should in general not need to store shared_ptr's globally.

**Comment by suxunbin** [comment]
However, I need thread scheduling. So I only store them in query_handler().
Another thread is responsible for query_thread().
I can't directly pass the shared_ptr to query_thread() in query_handler().
I need another thread to know all the shared_ptrs in order to assign.

**Comment by suxunbin** [comment]
For example,if there are 60 queries but I only have 30 threads in the threadpool. If I directly pass the shared_ptr to query_thread().Then there are 30 shared_ptr that can't pass.I must store them.
So how should I store the rest shared_ptr so that they are still valid when they leave the scope.

**Comment by eidheim** [comment]
Regarding a thread pool of size 30, here is a simplified example of how I would implement this using asio:

```c++
#include "server_http.hpp"

class Workers {
public:
  boost::asio::io_service service;

private:
  boost::asio::io_service::work work;
  std::vector<std::thread> threads;

public:
  Workers(size_t number_of_threads) : work(service) {
    for(size_t c = 0; c < number_of_threads; ++c) {
      threads.emplace_back([this] {
        service.run();
      });
    }
  }
};

using namespace std;

using HttpServer = SimpleWeb::Server<SimpleWeb::HTTP>;

int main() {
  Workers workers(30);

  HttpServer server;
  server.config.port = 8080;

  server.default_resource["GET"] = [&workers](shared_ptr<HttpServer::Response> response, shared_ptr<HttpServer::Request> /*request*/) {
    workers.service.post([response] {
      this_thread::sleep_for(chrono::seconds(5)); // Simulate time-consuming work
      response->write("Work done");
    });
  };

  server.start();
}
```
The response-shared_ptr is here copied by capture in `workers.service.post` and thus kept alive throughout the handler.

**Comment by suxunbin** [comment]
In this way,if 30 threads are running,then what about the new shared_ptr ?

**Comment by suxunbin** [comment]
If use workers.service.post, how should I sort these handlers in io_service.
I could not handle thest tasks according to the order of their entry.

**Comment by suxunbin** [comment]
![11](https://user-images.githubusercontent.com/22476651/39048905-c58d24ec-44d1-11e8-8ef6-745131988eb0.PNG)
![13](https://user-images.githubusercontent.com/22476651/39048909-ca178822-44d1-11e8-9006-ec08f237e775.PNG)
I use the example you give me. But client gets nothing.

**Comment by suxunbin** [comment]
![14](https://user-images.githubusercontent.com/22476651/39049028-3d5f5008-44d2-11e8-9480-38b378bd6ac5.PNG)
If I don't use workers.service.post(), client can directly get the data.

**Comment by suxunbin** [comment]
So, when I leave this scope, the response shared_ptr becomes invalid.
